clash-prelude/src/Clash/Class/Finite/Internal.hs

```diff
@@ -0,0 +1,751 @@
+import Clash.Class.Finite.Internal.Dictionaries
+  ( powGTZero, powMonotone1, powLawsRewrite
+#if !MIN_VERSION_base(4,16,0)
+  , powCLogDual
+#endif
+  )
+import Clash.Class.Finite.Internal.TH (deriveFiniteTuples)
+import Clash.Class.Resize (Resize(..))
+import Clash.Class.BitPack (BitPack(..), bitCoerce)
+import Clash.Promoted.Nat
+  ( SNat(..), UNat(..), toUNat, fromUNat, natToNum, snatToNum
+  )
+import Clash.Sized.Index (Index)
+import Clash.Sized.Internal.BitVector (BitVector(..), Bit(..), high, low)
+import Clash.Sized.Internal.Unsigned (Unsigned(..))
+import Clash.Sized.Signed (Signed)
+import Clash.Sized.RTree (RTree(..), tdfold, tfold, trepeat)
+import Clash.Sized.Vector
+  ( Vec(..), (++), (!!), concatMap, bv2v, dfold, foldl, foldr
+  , ifoldr, indicesI, iterateI, unfoldrI, repeat, reverse, replace
+  )
+import Clash.XException (ShowX, NFDataX)
+
+import qualified Data.List as List (iterate)
+
+{- $setup
+>>> :m -Prelude
+>>> :set -XDeriveAnyClass
+>>> import Clash.Prelude
+-}
+
+-- * Finite Class
+
+-- | The class of types holding only a finite number of elements. The
+-- class offers type level access to the number of elements @n@ and
+-- defines a total order on the elements via indexing them from @0@ to
+-- @n-1@. The class gives access to the vector of all elements of the
+-- type, allows to iterate over the elements in order or to map back
+-- and forth between their associated indices.
+--
+-- In general, the class can be considered as a more hardware-friendly
+-- alternative to 'Bounded' and 'Enum', utilizing 'Index' instead of
+-- 'Int' and vectors instead of lists.
+--
+-- The class supports generic deriving, i.e., for custom types class
+-- instances can be derived via @derive ('Generic', Finite)@ requiring
+-- that all inner types of the type declaration have @Finite@
+-- instances as well.
+--
+-- >>> data T = B Bit | D (Index 2) (Signed 1) deriving (Generic, Finite, Show)
+-- >>> natToNum @(ElementCount T)
+-- 6
+-- >>> elements @T
+-- B 0 :> B 1 :> D 0 -1 :> D 0 0 :> D 1 -1 :> D 1 0 :> Nil
+-- >>> lowest @T
+-- Just (B 0)
+-- >>> highest @T
+-- Just (D 1 0)
+-- >>> next (B 1)
+-- Just (D 0 -1)
+-- >>> prev (B 0)
+-- Nothing
+-- >>> index (D 0 0)
+-- 3
+-- >>> ith @T 5
+-- D 1 0
+--
+-- Any definition must satisfy the following laws (automatically
+-- ensured when generic deriving the instance):
+--
+-- [Index Order]
+--   @ index '<$>' elements = 'indicesI' @
+-- [Forward Iterate]
+--   @ 'iterateI' ('>>=' next) (lowest \@a) = 'Just' '<$>' (elements \@a) @
+-- [Backward Iterate]
+--   @ 'iterateI' ('>>=' prev) (highest \@a) = 'Just' '<$>' (elements \@a) @
+-- [Index Isomorphism]
+--   @ith (index x) = x@
+-- [Minimum Predecessor]
+--   @ lowest '>>=' prev = 'Nothing' @
+-- [Maximum Successor]
+--   @ highest '>>=' next = 'Nothing' @
+-- [No Uninhabited Extremes]
+--   @ lowest \@a = 'Nothing' /and/ highest \@a = 'Nothing'
+--     /if and only if/ ElementCount a = 0 @
```

Comment:
Would it be sensible to add hedgehog laws for this? Could these be added to the test suite for the types that derive / define an instance?

Reply:
Definitely! It's just not necessary at the current code reviewing stage IMO. Hence, I considered it to be a nice addition for a later stage. I also tested all of the features in REPL already :tm:.